readme.md
# simple-ifconfig

[![Build Status](https://travis-ci.org/brozeph/simple-ifconfig.svg?branch=master)](https://travis-ci.org/brozeph/simple-ifconfig) [![Coverage Status](https://coveralls.io/repos/github/brozeph/simple-ifconfig/badge.svg?branch=master)](https://coveralls.io/github/brozeph/simple-ifconfig?branch=master)

This module aims to provide additional information beyond what the `os.networkInterfaces()` method returns (via Node) for each detected network interface.

## Requirements

* `Node.js`: >= `v6.x`
* `Platform`: `Darwin`, `Unix` or `Linux` (_Windows is not supported at this time_)

## Installation

```bash
npm install simple-ifconfig
```

## Usage

### Constructor

Creates a new instance of the `NetworkInfo` class that allows for interaction with the networking interfaces of the system.

`new NetworkInfo(options)`

#### Options (optional)

* `ifconfigPath` (_optional_, `String`) - defaults to `/sbin/ifconfig` when not provided; defines the path to the `ifconfig` executable
* `active` (_optional_, `Boolean`) - defaults to `true` when not provided; when `true`, only the interfaces that are connected and actively transmitting data are included
* `internal` (_optional_, `Boolean`) - defaults to `false` when not provided; when `true`, internal (i.e. loopback, etc.) adapters are included when interacting with various other module methods
* `verbose` (_optional_, `Booelan`) - defaults to `true` when not provided; when `true`, `ifconfig` is called with the verbose flag (`-v`) - on some distros (i.e. `Alpine Linux`) this flag is not supported and it is helpful to turn it off

```javascript
import { NetworkInfo } from 'simple-ifconfig';

const
  ifconfigPath = '/sbin/ifconfg',
  active = false,
  internal = false;

let networking = new NetworkInfo({
  ifconfigPath,
  active,
  internal
});

// work with networking instance...
```

### #applySettings

Uses `ifconfig` to set `hardwareAddress`, `ipv4`, `mtu` as well as `up`/`down` (active) status for an interface.

`new NetworkInfo().applySettings(interfaceName, settings)`

```javascript
import { NetworkInfo } from 'simple-ifconfig';

let networking = new NetworkInfo();

networking
  .applySettings(`eth0`, {
    active : true,
    hardwareAddress : 'AA:BB:CC:00:11:22',
    ipv4 : {
      address : '192.168.0.2',
      broadcast : '192.168.0.1',
      netmask : '255.255.255.0'
    },
    mtu : 1800
  })
  .then(console.log)
  .catch(console.error);
```

* `active` (_optional_, `Boolean`) - brings the interface specified up when `true` or down when `false`
* `hardwareAddress` (_optional_, `String`) - will change the MAC address of the specified interface to the provided value when specified
* `ipv4` (_optional_, `Object`) - will allow for specifying IP address (STATIC) configuration for the specified interface
  * `address` (_optional_, `String`)
  * `broadcast` (_optional_, `String`)
  * `netmask` (_optional_, `String`)
* `mtu` (_optional_, `Number`) - when provided, sets the MTU for the specified interface

**note**: in order to clear a statically configured IP address, most OS variants support providing `0.0.0.0` to "unset" the previously specified `address`, `broadcast` and `subnet`

### #listInterfaces

Retrieves an array of available interfaces, optionally filtered according to [options](#options-optional) provided to the [constructor](#constructor).

`new NetworkInfo().listInterfaces()`

```javascript
import { NetworkInfo } from 'simple-ifconfig';

let networking = new NetworkInfo();

networking.listInterfaces()
  .then(console.log)
  .catch(console.error);
```

Example output:

```javascript
[ { hardwareAddress: '28:cf:e9:17:99:a9',
    internal: false,
    name: 'en0',
    flags:
     { broadcast: true,
       multicast: true,
       running: true,
       simplex: true,
       smart: true,
       up: true },
    index: 4,
    mtu: 1500,
    ipv6: [ { address: 'fe80::8f0:d3fc:39a2:bb9d', prefixLength: 64 } ],
    ipv4:
     [ { address: '10.129.14.60',
         netmask: '255.255.255.0',
         broadcast: '10.129.14.255' } ],
    active: true },
  { hardwareAddress: 'a8:20:66:16:d4:9d',
    internal: false,
    name: 'en3',
    flags:
     { broadcast: true,
       multicast: true,
       running: true,
       simplex: true,
       smart: true,
       up: true },
    index: 8,
    mtu: 1500,
    ipv6: [ { address: 'fe80::48e:ceba:32c2:923c', prefixLength: 64 } ],
    ipv4:
     [ { address: '10.129.41.23',
         netmask: '255.255.255.0',
         broadcast: '10.129.41.255' } ],
    active: true },
  { hardwareAddress: '4e:98:56:5b:36:97',
    internal: false,
    name: 'awdl0',
    flags:
     { broadcast: true,
       multicast: true,
       promisc: true,
       running: true,
       simplex: true,
       up: true },
    index: 11,
    mtu: 1484,
    ipv6: [ { address: 'fe80::4c98:56ff:fe5b:3697', prefixLength: 64 } ],
    active: true } ]
```

#### interface

Each interface returned is an object with the following properties:

* `active` (_required_, `Boolean`) - indicates whether or not the interface is connected and actively transmitting data
* `flags` (_optional_, `String`) - a sub-document that provides additional detail regarding the interface and its hardware configuration
  * `addrconf` (_optional_, `Boolean`)
  * `allmulti` (_optional_, `Boolean`)
  * `anycast` (_optional_, `Boolean`)
  * `broadcast` (_optional_, `Boolean`)
  * `cluif` (_optional_, `Boolean`)
  * `cos` (_optional_, `Boolean`)
  * `debug` (_optional_, `Boolean`)
  * `deprecated` (_optional_, `Boolean`)
  * `dhcp` (_optional_, `Boolean`)
  * `duplicate` (_optional_, `Boolean`)
  * `failed` (_optional_, `Boolean`)
  * `fixedmtu` (_optional_, `Boolean`)
  * `grouprt` (_optional_, `Boolean`)
  * `inactive` (_optional_, `Boolean`)
  * `loopback` (_optional_, `Boolean`)
  * `mip` (_optional_, `Boolean`)
  * `multibcast` (_optional_, `Boolean`)
  * `multicast` (_optional_, `Boolean`)
  * `multinet` (_optional_, `Boolean`)
  * `noarp` (_optional_, `Boolean`)
  * `nochecksum` (_optional_, `Boolean`)
  * `nofailover` (_optional_, `Boolean`)
  * `nolocal` (_optional_, `Boolean`)
  * `nonud` (_optional_, `Boolean`)
  * `nortexch` (_optional_, `Boolean`)
  * `notrailers` (_optional_, `Boolean`)
  * `noxmit` (_optional_, `Boolean`)
  * `oactive` (_optional_, `Boolean`)
  * `offline` (_optional_, `Boolean`)
  * `pfcopyall` (_optional_, `Boolean`)
  * `pointopoint` (_optional_, `Boolean`)
  * `preferred` (_optional_, `Boolean`)
  * `private` (_optional_, `Boolean`)
  * `pseg` (_optional_, `Boolean`)
  * `promisc` (_optional_, `Boolean`)
  * `quorumloss` (_optional_, `Boolean`)
  * `router` (_optional_, `Boolean`)
  * `running` (_optional_, `Boolean`)
  * `simplex` (_optional_, `Boolean`)
  * `smart` (_optional_, `Boolean`)
  * `standby` (_optional_, `Boolean`)
  * `temporary` (_optional_, `Boolean`)
  * `unnumbered` (_optional_, `Boolean`)
  * `up` (_optional_, `Boolean`)
  * `virtual` (_optional_, `Boolean`)
  * `varmtu` (_optional_, `Boolean`)
  * `xresolv` (_optional_, `Boolean`)
* `hardwareAddress` (_required_, `String`) - the MAC address assigned to the interface
* `index` (_optional_, `String`) - on `Darwin` OS, this is functionality equivalent to `metric`... it is the prioritized order of the interface to the OS
* `internal` (_required_, `Boolean`) - indicates whether the interface exists for internal use within the OS (i.e. a loopback interface)
* `ipv4` (_optional_, `Array`) - IPv4 address information for the interface
  * `address` (_optional_, `String`)
  * `broadcast` (_optional_, `String`)
  * `netmask` (_optional_, `String`)
* `ipv6` (_optional_, `Array`) - IPv6 address information for the interface
  * `address` (_optional_, `String`)
  * `prefixLength` (_optional_, `Number`)
* `metric` (_optional_, `Number`) - the prioritized order of the interface to the OS
* `mtu` (_optional_, `Number`) - the maximum transmission unit size for the interface
* `name` (_required_, `String`) - the name assigned to the interface


history.md
# v0.2.3 - 10/06/2020

* Updated dependencies

# v0.2.2 - 02/08/2019

* Updated gulp and Babel dependencies

# v0.2.1 - 09/02/2017

* Added configuration to disable `verbose` for use on certain distros of Linux (i.e. Alpine)

# v0.2.0 - 08/24/2017

* Added `applySettings` command

# v0.1.2 - 04/02/2017

* Fixed issue where Linux status was not always properly parsed

# v0.1.1 - 12/08/2016

* Updated documentation

# v0.1.0 - 12/08/2016

* Initial release

test/example.js
const ifconfig = require('../dist');

module.export = (function () {
	'use strict';

	let network = new ifconfig.NetworkInfo();

	network
		.listInterfaces()
		.then(console.log)
		.catch(console.error);
}());


gulpfile.js
'use strict';

var
	babel = require('gulp-babel'),
	coveralls = require('gulp-coveralls'),
	del = require('del'),
	eslint = require('gulp-eslint'),
	gulp = require('gulp'),
	gulpUtil = require('gulp-util'),
	istanbul = require('gulp-istanbul'),
	mocha = require('gulp-mocha'),
	sourcemaps = require('gulp-sourcemaps');


gulp.task('build', ['clean'], () => {
	return gulp
		.src('src/**/*.js')
		.pipe(sourcemaps.init())
		.pipe(babel({
			presets : ['es2015', 'stage-0']
		}))
		.pipe(sourcemaps.write('.'))
		.pipe(gulp.dest('dist'));
});


gulp.task('clean', function (callback) {
	return del(['dist', 'reports'], callback);
});


gulp.task('coveralls', ['test-coverage'], function () {
	return gulp
		.src('reports/lcov.info')
		.pipe(coveralls());
});


gulp.task('lint', function () {
	return gulp
		.src(['**/*.js', '!dist/**', '!node_modules/**', '!reports/**'])
		.pipe(eslint())
		.pipe(eslint.format())
		.pipe(eslint.failAfterError());
});


gulp.task('test-coverage', ['build'], function () {
	return gulp
		.src(['./dist/**/*.js'])
		.pipe(istanbul())
		.pipe(istanbul.hookRequire())
		.on('finish', function () {
			gulp
				.src(['./test/**/*.js'])
				.pipe(mocha({ reporter : 'spec' })
						.on('error', function (err) {
							if (err.showStack) {
								gulpUtil.log(err);
							}

							/*eslint no-invalid-this:0*/
							this.emit('end');
						}))
				.pipe(istanbul.writeReports('./reports'));
		});
});


gulp.task('test-unit', ['build'], function () {
	return gulp
		.src(['./test/**/*.js'], { read : false })
		.pipe(mocha({
			checkLeaks : true,
			reporter : 'spec',
			ui : 'bdd'
		}));
});


src/index.js
import logger from 'debug';
import { spawn } from 'child_process';

const
	BASE_16 = 16,
	CHAR_ADVANCE = 2,
	debug = logger('simple-ifconfig'),
	DEFAULT_ACTIVE_DOWN = 'down',
	DEFAULT_ACTIVE_UP = 'up',
	DEFAULT_HARDWARE_ADDR = '00:00:00:00:00:00',
	DEFAULT_METRIC = 99,
	DEFAULT_OPTIONS = {
		active : true,
		ifconfigPath : '/sbin/ifconfig',
		internal : false,
		verbose : true
	},
	RE_DELIM = /\ |\:/,
	// http://www-01.ibm.com/support/docview.wss?uid=isg3T1019709
	// http://docs.oracle.com/cd/E19253-01/816-5166/6mbb1kq31/#INTERFACE%20FLAGS
	RE_FLAGS = {
		addrconf : /(^|[\ \t\,]*)addrconf($|[\ \t\,]*)/i,
		allmulti : /(^|[\ \t\,]*)allmulti($|[\ \t\,]*)/i,
		anycast : /(^|[\ \t\,]*)anycast($|[\ \t\,]*)/i,
		broadcast : /(^|[\ \t\,]*)broadcast($|[\ \t\,]*)/i,
		cluif : /(^|[\ \t\,]*)cluif($|[\ \t\,]*)/i,
		cos : /(^|[\ \t\,]*)cos($|[\ \t\,]*)/i,
		debug : /(^|[\ \t\,]*)debug($|[\ \t\,]*)/i,
		deprecated : /(^|[\ \t\,]*)deprecated($|[\ \t\,]*)/i,
		dhcp : /(^|[\ \t\,]*)dhcp($|[\ \t\,]*)/i,
		duplicate : /(^|[\ \t\,]*)duplicate($|[\ \t\,]*)/i,
		failed : /(^|[\ \t\,]*)failed($|[\ \t\,]*)/i,
		fixedmtu : /(^|[\ \t\,]*)fixedmtu($|[\ \t\,]*)/i,
		grouprt : /(^|[\ \t\,]*)grouprt($|[\ \t\,]*)/i,
		inactive : /(^|[\ \t\,]*)inactive($|[\ \t\,]*)/i,
		loopback : /(^|[\ \t\,]*)loopback($|[\ \t\,]*)/i,
		mip : /(^|[\ \t\,]*)mip($|[\ \t\,]*)/i,
		multibcast : /(^|[\ \t\,]*)multi_bcast($|[\ \t\,]*)/i,
		multicast : /(^|[\ \t\,]*)multicast($|[\ \t\,]*)/i,
		multinet : /(^|[\ \t\,]*)multinet($|[\ \t\,]*)/i,
		noarp : /(^|[\ \t\,]*)noarp($|[\ \t\,]*)/i,
		nochecksum : /(^|[\ \t\,]*)nochecksum($|[\ \t\,]*)/i,
		nofailover : /(^|[\ \t\,]*)nofailover($|[\ \t\,]*)/i,
		nolocal : /(^|[\ \t\,]*)nolocal($|[\ \t\,]*)/i,
		nonud : /(^|[\ \t\,]*)nonud($|[\ \t\,]*)/i,
		nortexch : /(^|[\ \t\,]*)notexch($|[\ \t\,]*)/i,
		notrailers : /(^|[\ \t\,]*)notrailers($|[\ \t\,]*)/i,
		noxmit : /(^|[\ \t\,]*)noxmit($|[\ \t\,]*)/i,
		oactive : /(^|[\ \t\,]*)oactive($|[\ \t\,]*)/i,
		offline : /(^|[\ \t\,]*)offline($|[\ \t\,]*)/i,
		pfcopyall : /(^|[\ \t\,]*)pfcopyall($|[\ \t\,]*)/i,
		pointopoint : /(^|[\ \t\,]*)pointopoint($|[\ \t\,]*)/i,
		preferred : /(^|[\ \t\,]*)preferred($|[\ \t\,]*)/i,
		private : /(^|[\ \t\,]*)private($|[\ \t\,]*)/i,
		promisc : /(^|[\ \t\,]*)promisc($|[\ \t\,]*)/i,
		pseg : /(^|[\ \t\,]*)pseg($|[\ \t\,]*)/i,
		quorumloss : /(^|[\ \t\,]*)quorumloss($|[\ \t\,]*)/i,
		router : /(^|[\ \t\,]*)router($|[\ \t\,]*)/i,
		running : /(^|[\ \t\,]*)running($|[\ \t\,]*)/i,
		simplex : /(^|[\ \t\,]*)simplex($|[\ \t\,]*)/i,
		smart : /(^|[\ \t\,]*)smart($|[\ \t\,]*)/i,
		standby : /(^|[\ \t\,]*)standby($|[\ \t\,]*)/i,
		temporary : /(^|[\ \t\,]*)temporary($|[\ \t\,]*)/i,
		unnumbered : /(^|[\ \t\,]*)unnumbered($|[\ \t\,]*)/i,
		up : /(^|[\ \t\,]*)up($|[\ \t\,]*)/i,
		varmtu : /(^|[\ \t\,]*)var\_mtu($|[\ \t\,]*)/i,
		virtual : /(^|[\ \t\,]*)virtual($|[\ \t\,]*)/i,
		xresolv : /(^|[\ \t\,]*)xresolv($|[\ \t\,]*)/i
	},
	RE_HARDWARE_ADDR = /(ether|hwaddr)\ +(([0-9a-f]{2}[\:\-]{0,1}){6})/i,
	RE_IFCONFIG_FLAGS = /<?([a-z\,\ \t\_]*)\>?(([\ \t]*mtu[\:\ \t]+[0-9]+)|([\ \t]*metric[\:\ \t]+[0-9]+)|([\ \t]*index[\:\ \t]+[0-9]+))+/i,
	RE_IFCONFIG_IPV4 = /^\s*inet\s/,
	RE_IFCONFIG_IPV6 = /^\s*inet6\s/,
	RE_INDEX = /index[\ \:]+[0-9]+/i,
	RE_LINUX_ADDR = /^addr\:/i,
	RE_LINUX_BCAST = /^bcast\:/i,
	RE_LINUX_MASK = /^mask\:/i,
	RE_METRIC = /metric[\ \:]+[0-9]+/i,
	RE_MTU = /mtu[\ \:]+[0-9]+/i,
	RE_NUMBER = /^[0-9]{2,}/,
	RE_STATUS = /^\s*(status)|((r|t)x\ bytes)/i,
	RE_UNIX_ADDR = /^inet$/i,
	RE_UNIX_BCAST = /^broadcast$/i,
	RE_UNIX_IPV6_ADDR = /^inet6$/i,
	RE_UNIX_IPV6_PREFIX_LENGTH = /^prefixlen$/i,
	RE_UNIX_MASK = /^netmask$/i,
	RE_UNIX_STATUS_ACTIVE = /\ active$/i,
	VERBOSE = '-v';

function _ensureDefaultOptions (networkInfo) {
	Object
		.getOwnPropertyNames(DEFAULT_OPTIONS)
		.forEach((optionName) => {
			debug('examining option %s (%s)', optionName, networkInfo.options[optionName]);
			if (_isNullOrUndefined(networkInfo.options[optionName])) {
				debug(
					'applying default value for option %s (%o)',
					optionName,
					DEFAULT_OPTIONS[optionName]);

				networkInfo.options[optionName] = DEFAULT_OPTIONS[optionName];
			}
		});
}

function _ifconfig (networkInfo, ...args) {
	return new Promise((resolve, reject) => {
		let
			ifconfig = spawn(networkInfo.options.ifconfigPath, args),
			stderr = [],
			stdout = [];

		// handle command exit
		ifconfig.on('close', (code) => {
			debug(
				'%s %s command completed (code: %s)',
				networkInfo.options.ifconfigPath,
				args.join(' '),
				code);

			if (code) {
				let err = new Error(stderr.join(''));
				err.code = code;
				err.command = [networkInfo.options.ifconfigPath].concat(args).join(' ');

				return reject(err);
			}

			return resolve(stdout.join(''));
		});

		// handle errors while attempting to execute command
		ifconfig.on('error', (err) => {
			err.command = [networkInfo.options.ifconfigPath].concat(args).join(' ');
			debug('%s command failed (error: %s)', err.command, err.message);

			return reject(err);
		});

		// capture command output
		ifconfig.stdout.on('data', (chunk) => stdout.push(chunk));
		ifconfig.stderr.on('data', (chunk) => stderr.push(chunk));
	});
}

function _isNullOrUndefined (value) {
	return value === null || typeof value === 'undefined';
}

function _parseInterfaceInfo (networkInfo, ifconfigResult) {
	let
		addr,
		hardwareInterfaces = new Map(),
		iface,
		result = [];

	// break the ifconfig command result into lines and parse them 1 by 1...
	ifconfigResult.split(/\r?\n/).forEach((line) => {
		let terms;

		// look for a new interface line
		if (!/\s/.test(line.charAt(0))) {
			// split the line on spaces (and optionally a colon)
			line = line.split(/\:?\ +/);

			// create new iface...
			iface = {
				active : false,
				hardwareAddress : DEFAULT_HARDWARE_ADDR,
				internal : true
			};

			// assign the name using the first term in the line
			iface.name = line[0];

			// ensure we have a valid interface
			if (!iface.name) {
				return;
			}

			// track the interface to return after processing
			hardwareInterfaces.set(iface.name, iface);

			// rebuild the line minus the adapter name and continue processing
			line = line.slice(1).join(' ');
		}

		// look for hardware address
		if (RE_HARDWARE_ADDR.test(line)) {
			terms = line.match(RE_HARDWARE_ADDR);
			debug(
				'hardware address of %s found for interface %s',
				terms[2],
				iface.name);
			iface.hardwareAddress = terms[2];
			iface.internal = false;
		}

		// look for flag information and process
		if (RE_IFCONFIG_FLAGS.test(line)) {
			debug('interface flags found for interface %s (%s)', iface.name, line);
			iface.flags = {};

			// map flags to the interface
			Object.keys(RE_FLAGS).forEach((flagName) => {
				if (RE_FLAGS[flagName].test(line)) {
					debug('flag %s found for interface %s', flagName, iface.name);
					iface.flags[flagName] = true;
				}
			});

			if (RE_INDEX.test(line)) {
				terms = line.match(RE_INDEX)[0].split(RE_DELIM);
				debug('index of %d found for interface %s', terms[1], iface.name);
				iface.index = parseInt(terms[1], 10);
			}

			if (RE_METRIC.test(line)) {
				terms = line.match(RE_METRIC)[0].split(RE_DELIM);
				debug('metric of %d found for interface %s', terms[1], iface.name);
				iface.metric = parseInt(terms[1], 10);
			}

			if (RE_MTU.test(line)) {
				terms = line.match(RE_MTU)[0].split(RE_DELIM);
				debug('mtu of %d found for interface %s', terms[1], iface.name);
				iface.mtu = parseInt(terms[1], 10);
			}

			return;
		}

		// look for IPv4 info...
		if (RE_IFCONFIG_IPV4.test(line)) {
			debug('IPv4 information found for interface %s (%s)', iface.name, line);

			addr = {};
			iface.ipv4 = (iface.ipv4 || []).concat([addr]);
			terms = line.split(/\s+/);

			terms.forEach((term, i) => {
				// linux formatting - addr:10.0.2.15
				if (RE_LINUX_ADDR.test(term)) {
					addr.address = term.split(RE_LINUX_ADDR)[1];
					return;
				}

				// linux formatting - Bcast:10.0.2.255
				if (RE_LINUX_BCAST.test(term)) {
					addr.broadcast = term.split(RE_LINUX_BCAST)[1];
					return;
				}

				// linux formatting - Mask:255.255.255.0
				if (RE_LINUX_MASK.test(term)) {
					addr.netmask = term.split(RE_LINUX_MASK)[1];
					return;
				}

				// unix formatting - address is 1st term
				if (RE_UNIX_ADDR.test(term)) {
					addr.address = terms[i + 1];
					return;
				}

				// unix formatting - netmask 0xffffff00
				if (RE_UNIX_MASK.test(term)) {
					let netmask = terms[i + 1];

					// convert from hexidecimal if applicable
					if (/^0x/.test(netmask)) {
						// iterate 2 chars at a time converting from hex to decimal
						for (let i = CHAR_ADVANCE; i < netmask.length; i += CHAR_ADVANCE) {
							addr.netmask = [
								addr.netmask || '',
								addr.netmask ? '.' : '',
								parseInt(
									['0x', netmask.slice(i, i + CHAR_ADVANCE)].join(''),
									BASE_16)].join('');
						}

						return;
					}

					// set netmask in teh event it is not in hexidecimal format
					addr.netmask = netmask;
					return;
				}

				// unix formatting - broadcast 10.129.8.255
				if (RE_UNIX_BCAST.test(term)) {
					addr.broadcast = terms[i + 1];
				}
			});
		}

		// look for IPv6 info
		if (RE_IFCONFIG_IPV6.test(line)) {
			debug('IPv6 information found for interface %s (%s)', iface.name, line);

			addr = {};
			iface.ipv6 = (iface.ipv6 || []).concat([addr]);
			terms = line.split(/\ +/);

			terms.some((term, i) => {
				// check for linux address reference
				if (RE_LINUX_ADDR.test(term)) {
					terms = terms[i + 1].split(/\//);
					addr.address = terms[0];
					addr.prefixLength = parseInt(terms[1], 10);

					return true;
				}

				if (RE_UNIX_IPV6_ADDR.test(term.trim())) {
					addr.address = terms[i + 1].split(/\%/)[0];
				}

				if (RE_UNIX_IPV6_PREFIX_LENGTH.test(term)) {
					addr.prefixLength = parseInt(terms[i + 1], 10);

					return true;
				}

				return false;
			});
		}

		// look for status
		if (RE_STATUS.test(line)) {
			debug('status information found for interface %s (%s)', iface.name, line);

			terms = line.split(/\:/g);
			iface.active =
				RE_UNIX_STATUS_ACTIVE.test(terms[1]) ||
				RE_NUMBER.test(terms[1]);
		}
	});

	// populate the internal interfaces array
	hardwareInterfaces.forEach((iface) => {
		let include =
			// filter out internal interfaces as applicable
			(networkInfo.options.internal || !iface.internal) &&
			// filter out non-active interfaces as applicable
			(!networkInfo.options.active || iface.active);

		if (include) {
			result.push(iface);
		}
	});

	// sort by metric / priority
	result.sort((a, b) => (
		(a.metric || a.index || DEFAULT_METRIC) - (b.metric || b.index || DEFAULT_METRIC)));

	return result;
}

export class NetworkInfo {
	constructor (options) {
		debug('new NetworkInfo(%o)', options);

		this._options = options || {};

		_ensureDefaultOptions(this);
	}

	async applySettings (name, settings) {
		if (_isNullOrUndefined(name)) {
			throw new Error('interface name is required');
		}

		if (_isNullOrUndefined(settings)) {
			throw new Error('settings are required');
		}

		// get the interface... if it does not exist, this will result in error
		let
			iface = await _ifconfig(this, name),
			result;

		// check to see if adapter should be enabled / disabled
		if (!_isNullOrUndefined(settings.active)) {
			debug(
				'attempt %sabling interface %s',
				settings.active ? 'en' : 'dis',
				name);

			if (settings.active) {
				result = await _ifconfig(this, name, DEFAULT_ACTIVE_UP);
			} else {
				result = await _ifconfig(this, name, DEFAULT_ACTIVE_DOWN);
			}

			debug(
				'result of %sabling %s: %o',
				settings.active ? 'en' : 'dis',
				name,
				result);
		}

		// check for hardware address update
		if (!_isNullOrUndefined(settings.hardwareAddress)) {
			debug('attempt setting hardware address for interface %s', name);

			result = await _ifconfig(
				this,
				name,
				'hw',
				'ether',
				settings.hardwareAddress);

			debug(
			'result of setting hardware address for interface %s: %o',
			name,
			result);
		}

		// check for ipv4 settings updates
		if (!_isNullOrUndefined(settings.ipv4)) {
			let
				addr = settings.ipv4,
				args = [name];

			// if an array, select the first object
			if (Array.isArray(addr)) {
				addr = addr[0];
			}

			if (!_isNullOrUndefined(addr.address)) {
				debug('applying ipv4 hardware address: %s', addr.address);
				args.push(addr.address);
			}

			if (!_isNullOrUndefined(addr.broadcast)) {
				debug('applying ipv4 hardware broadcast: %s', addr.broadcast);
				args.push('broadcast', addr.broadcast);
			}

			if (!_isNullOrUndefined(addr.netmask)) {
				debug('applying ipv4 hardware netmask: %s', addr.netmask);
				args.push('netmask', addr.netmask);
			}

			if (args.length > 1) {
				result = await _ifconfig(this, ...args);

				debug(
					'result of ipv4 settings update to %s: %o',
					name,
					result);
			}
		}

		// check for MTU update
		if (!_isNullOrUndefined(settings.mtu)) {
			debug('attempt setting mtu for interface %s', name);

			result = await _ifconfig(this, name, 'mtu', settings.mtu);

			debug(
				'result of setting mtu for interface %s: %o',
				name,
				result);
		}

		// reload interface settings after applying them...
		iface = await _ifconfig(this, name);

		return _parseInterfaceInfo(this, iface)[0];
	}

	async listInterfaces () {
		let result = this.options.verbose ?
			await _ifconfig(this, VERBOSE) :
			await _ifconfig(this);

		return _parseInterfaceInfo(this, result);
	}

	get options () {
		return this._options;
	}
}

export default { NetworkInfo };
